### PR TITLE
Add input

### PR DESCRIPTION
**Pull request by tobiaslins**

### CODE REVIEW

File: web/src/components/text-input/search-field.tsx [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+  return (
+    <GenericField
+      {...props}
+      component="input"
+      renderLeft={() => (
+        <div className={classes.leftIcon}>
+          <Icon name="search" size="small" />
+        </div>
+      )}
+      renderRight={({ onClear, dirty }) => {
```
Comment (AleshaOleg) [inline web/src/components/text-input/search-field.tsx:38]: `isDirty` please :)

File: web/src/components/text-input/generic.tsx [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+      )}
+      {validationHelp && (
+        <Flex gap={1}>
+          <Icon name="alert" color="danger" size="small" />
+          <Text as="span" size="body-small" color="danger">
+            {validationHelp}
+          </Text>
+        </Flex>
+      )}
+      <div style={{ position: 'relative' }}>
```
Comment (AleshaOleg) [inline web/src/components/text-input/generic.tsx:100]: Move this to `styles.ts`

File: web/src/components/text-input/generic.tsx [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+        {renderLeft?.()}
+
+        <Component
+          className={clsx(classes.textField, onFocus)}
+          {...inputProps}
+          {...((invalid || allowedChars < 0) && { 'aria-invalid': true })}
```
Comment (AleshaOleg) [inline web/src/components/text-input/generic.tsx:106]: But invalid already contain check for `allowedChars < 0`. Shouldn't this check be removed here? Or use `isInvalid` here and below and remove `invalid`.

File: web/package.json [hunk 4]
```diff
@@ -21,12 +21,13 @@
...
     "@react-aria/checkbox": "^3.2.1",
     "@react-aria/focus": "^3.2.1",
+    "@react-aria/textfield": "^3.2.1",
     "@react-aria/switch": "^3.1.1",
     "@react-aria/visually-hidden": "^3.2.1",
     "@react-stately/toggle": "^3.2.1",
     "@react-types/shared": "^3.2.0",
     "clsx": "^1.1.1",
-    "grape-icons": "https://github.com/ubergrape/grape-icons.git",
+    "grape-icons": "ubergrape/grape-icons",
```
Comment (AleshaOleg) [inline web/package.json:30]: should we publish this package?

File: web/src/components/checkbox/styles.ts [hunk 5]
```diff
@@ -2,37 +2,45 @@ import { createUseStyles } from 'react-jss'
...
+    return isChecked
       ? tokens.colorBorderFormcontrolOnDisabled
       : tokens.colorBorderFormcontrolOffDisabled
   }
-  if (invalid) return tokens.colorBorderFormcontrolInvalidDefault
-  if (checked) return tokens.colorBorderFormcontrolOnDefault
-  if (!checked && indeterminate) return tokens.colorBorderFormcontrolOnDefault
+  if (isInvalid) return tokens.colorBorderFormcontrolInvalidDefault
+  if (isChecked) return tokens.colorBorderFormcontrolOnDefault
+  if (!isChecked && isIndeterminate)
```
Comment (AleshaOleg) [inline web/src/components/checkbox/styles.ts:25]: Next time please do in separate PR, will be easier to review :)

File: web/src/components/text-input/styles.ts [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+    color: tokens.colorTextPrimary,
+    resize: 'none',
+    backgroundColor: ({ isDisabled, isReadOnly }) => {
+      if (isDisabled) return tokens.colorBackgroundSecondary
+      if (isReadOnly) return tokens.colorBackgroundFormfieldReadonly
+      return undefined
+    },
+    cursor: ({ isDisabled }) => {
+      if (isDisabled) return 'not-allowed'
+      return undefined
```
Comment (AleshaOleg) [inline web/src/components/text-input/styles.ts:25]: And `default` here?

File: web/src/components/text-input/styles.ts [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+        : tokens.borderRadiusFormfieldTextarea
+    },
+    borderWidth: tokens.borderWidthFormfield,
+    borderStyle: 'solid',
+    color: tokens.colorTextPrimary,
+    resize: 'none',
+    backgroundColor: ({ isDisabled, isReadOnly }) => {
+      if (isDisabled) return tokens.colorBackgroundSecondary
+      if (isReadOnly) return tokens.colorBackgroundFormfieldReadonly
+      return undefined
```
Comment (AleshaOleg) [inline web/src/components/text-input/styles.ts:21]: Maybe better to return `white` here?

File: web/src/components/text-input/styles.ts [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+    },
+    cursor: ({ isDisabled }) => {
+      if (isDisabled) return 'not-allowed'
+      return undefined
+    },
+    height: ({ component, maxLength }) => {
+      if (component === 'input') return tokens.sizeFormfieldInput
+      if (maxLength) {
+        return '92px'
```
Comment (AleshaOleg) [inline web/src/components/text-input/styles.ts:30]: Probably a new token should be added for this? What do you think? Or it's just single time usage?

File: web/src/components/text-input/styles.ts [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+    },
+    boxSizing: 'border-box',
+    padding: () => {
+      return `${tokens.paddingFormfieldTextareaInputtextTopbottom} ${tokens.paddingFormfieldInputtextLeftright}`
+    },
+    paddingRight: ({ maxLength, component }) => {
+      if (maxLength && component === 'input')
+        return `calc(${tokens.paddingFormfieldInputtextCounter} + 50px)`
+
+      return null
```
Comment (AleshaOleg) [inline web/src/components/text-input/styles.ts:42]: Probably should be replaced with default 0 here. Just will look more clear as I think

File: web/src/components/text-input/search-field.tsx [hunk 10]
```diff
@@ -0,0 +1,56 @@
+/* eslint-disable jsx-a11y/label-has-associated-control */
```
Comment (AleshaOleg) [inline web/src/components/text-input/search-field.tsx:1]: Is it covered by `aria`, correct? We can disable it then I think in general config

File: web/src/styles/scale.ts [hunk 11]
```diff
@@ -0,0 +1,26 @@
+import tokens from '../tokens/tokens.json'
+
+const {
```
Comment (AleshaOleg) [inline web/src/styles/scale.ts:3]: I think I'll use this in the `Space` component instead of a string value. Thanks!

**Review by AleshaOleg — CHANGES_REQUESTED** [review]

File: web/src/styles/global.tsx [hunk 12]
```diff
@@ -1,12 +1,29 @@
 import { createUseStyles } from 'react-jss'
 import tokens from '../tokens'
 
+const focusStyle = ({ isInvalid }): string => {
```
Comment (AleshaOleg) [inline web/src/styles/global.tsx:4]: Should we use `outline` here instead of `boxShadow`?

File: web/src/components/text-input/generic.tsx (hunk 3 above)
Comment (tobiaslins) [reply web/src/components/text-input/generic.tsx:106]: good catch thanks!

File: web/src/components/text-input/styles.ts (hunk 8 above)
Comment (tobiaslins) [reply web/src/components/text-input/styles.ts:30]: I think this is single use only. We can improve/add one if we have this value again

File: web/src/components/text-input/styles.ts (hunk 9 above)
Comment (tobiaslins) [reply web/src/components/text-input/styles.ts:42]: If we return null/undefined, no value will be set for `paddingLeft`/`paddingRight` that's what we want because `padding` is set and would overwrite the values

File: web/src/components/text-input/search-field.tsx (hunk 10 above)
Comment (tobiaslins) [reply web/src/components/text-input/search-field.tsx:1]: The label has an associated control but it doesn't validate it because of spreading the props and not explicit setting it

File: web/src/styles/global.tsx (hunk 12 above)
Comment (tobiaslins) [reply web/src/styles/global.tsx:4]: outlines look strange if border-radius is set. using `boxShadow` as focus style is very common. check out the input focus from github :)

Comment (AleshaOleg) [reply web/src/styles/global.tsx:4]: I see, well then I'll switch to `boxShadow` too

File: web/src/components/text-input/styles.ts (hunk 9 above)
Comment (AleshaOleg) [reply web/src/components/text-input/styles.ts:42]: Thanks for explanation, missed that

File: web/src/components/text-input/styles.ts (hunk 7 above)
Comment (pseudonymcp) [reply web/src/components/text-input/styles.ts]: Should we return the token for the default formfield background?

Comment (pseudonymcp) [reply web/src/components/text-input/styles.ts]: Also for disabled we should use the specified token. I just noticed a change in the design tokens to be made (change disabled-hgost from "gray-5" to "gray-2"). I'll do this then to look won't change and you can use the token.

**Review by pseudonymcp — CHANGES_REQUESTED** [review]
For text field, text area and search field:  
-  I found an issue with the label, help and validation text: The label and help/validation text should expand and be bound to the width of the field and could span multiple lines.  

For text area:  
- The read-only state is missing the correct style.  

For search field:  
- Icon color incorrect: prefx icon and icon in clear button are too dark -> there is a separate token for icons in formfields to use `color-background-formfield-icon`
- Is there a reason why you used `type=text` instead of `type=search`?

**Comment by tobiaslins** [comment]
> For text field, text area and search field:
> 
> * I found an issue with the label, help and validation text: The label and help/validation text should expand and be bound to the width of the field and could span multiple lines.
> 
> For text area:
> 
> * The read-only state is missing the correct style.
> 
> For search field:
> 
> * Icon color incorrect: prefx icon and icon in clear button are too dark -> there is a separate token for icons in formfields to use `color-background-formfield-icon`
> * Is there a reason why you used `type=text` instead of `type=search`?

done

**Review by pseudonymcp — CHANGES_REQUESTED** [review]
Looking at the functional part, what seems to be missing from the requirements is:  

- All user-facing text must be translatable (including "required" and "optional"), including text to be read by assistive technology.
- A required field should be able to be defined to show or hide the "required" denotation.

Text Field:  
- Possibility to define a maximum and/or a minimum value that is valid is missing. Values not matching set the field to invalid. So for example if the max. value is set to "10" and the user enters "12", it's invalid.  

Text Area:  
- A text area can be set to have a fixed height (= minimum height) or to auto-expand vertically to fit the text until a defined max. height (rows).
- It can be set to be vertically (not horizontally) resizable (default: no resizing).

Search Field:  
- The prefix icon can be defined and taken from all icons available in grape-icons via the Icon component. It needs to be customizable for different usages of the search field component.

We need to discuss if we want to use a container for every single field to define the size as individual fields inside forms will vary in length corresponding to the expected text input. @AleshaOleg What are your thoughts on this?

**Comment by tobiaslins** [comment]
> Looking at the functional part, what seems to be missing from the requirements is:
> 
> * All user-facing text must be translatable (including "required" and "optional"), including text to be read by assistive technology. 

✅

> * A required field should be able to be defined to show or hide the "required" denotation.
> 
> Text Field:
> 
> * Possibility to define a maximum and/or a minimum value that is valid is missing. Values not matching set the field to invalid. So for example if the max. value is set to "10" and the user enters "12", it's invalid.

✅

> 
> Text Area:
> 
> * A text area can be set to have a fixed height (= minimum height) or to auto-expand vertically to fit the text until a defined max. height (rows).
> * It can be set to be vertically (not horizontally) resizable (default: no resizing).
`minHeight` and `maxHeight` can be set. `allowResize` can be set

> 
> Search Field:
> 
> * The prefix icon can be defined and taken from all icons available in grape-icons via the Icon component. It needs to be customizable for different usages of the search field component.

✅
> 
> We need to discuss if we want to use a container for every single field to define the size as individual fields inside forms will vary in length corresponding to the expected text input. @AleshaOleg What are your thoughts on this?

**Review by pseudonymcp — APPROVED** [review]

File: web/src/components/text-input/text-field.tsx [hunk 13]
```diff
@@ -0,0 +1,11 @@
+import React from 'react'
+
+import { GenericField, InputProps } from './generic'
+
+export const TextField: React.FC<InputProps> = props => {
+  return <GenericField {...props} component="input" />
+}
+
+TextField.defaultProps = {}
```
Comment (AleshaOleg) [inline web/src/components/text-input/text-field.tsx:9]: Let's remove all of the empty defaultProps

File: web/src/components/text-input/styles.ts [hunk 14]
```diff
@@ -0,0 +1,91 @@
...
+      if (component === 'input') return tokens.sizeFormfieldInput
+      if (maxLength) {
+        return '92px'
+      }
+      return '60px'
+    },
+    minHeight: ({ minHeight }) => minHeight,
+    maxHeight: ({ maxHeight }) => maxHeight,
+    boxSizing: 'border-box',
+    padding: () => {
```
Comment (AleshaOleg) [inline web/src/components/text-input/styles.ts:38]: here can be a string instead of a function

**Review by AleshaOleg — CHANGES_REQUESTED** [review]

File: web/src/components/text-input/styles.ts (hunk 14 above)
Comment (tobiaslins) [reply web/src/components/text-input/styles.ts]: done

File: web/src/components/text-input/text-field.tsx (hunk 13 above)
Comment (tobiaslins) [reply web/src/components/text-input/text-field.tsx]: done

**Review by AleshaOleg — APPROVED** [review]
